cmd/gardenlet/app/app.go

```diff
@@ -460,6 +467,30 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
 	return flow.Parallel(taskFns...)(ctx)
 }
 
+// TODO(Kostov6): Remove this code after v1.89 has been released.
+func (g *garden) cleanupGRMSecretFinalizers(ctx context.Context, seedClient client.Client, log logr.Logger) error {
+	mrs := &resourcesv1alpha1.ManagedResourceList{}
+	if err := seedClient.List(ctx, mrs); err != nil {
+		log.Error(err, "Failed to list ManagedResources while cleaing up GRM finalizers")
+	}
+	secret := &corev1.Secret{}
+	for _, mr := range mrs.Items {
+		for _, ref := range mr.Spec.SecretRefs {
+			if err := g.mgr.GetClient().Get(ctx, client.ObjectKey{Namespace: mr.Namespace, Name: ref.Name}, secret); err != nil {
+				log.Error(err, "Failed to get secret while cleaing up GRM finalizers")
```

Comment:
return error instead of logging

Reply:
I wasn't sure how to handle the error and decided to log the error instead of throwing it because if the cleanup logic fails it will put gardenlet in CrashLoopBackOff and prevent doing it's intended work. Can you explain why returning the error is the preferred way because I don't well understand when to use one or the other?

Reply:
Returning the error is always the preferred way. There are just a few exceptional cases when this is not done, and such cases are those that should be justified.
In this specific case, an error will probably only happen when the seed's API server is down or has problems. Then `gardenlet` will have other issues anyways and will fail eventually.